cpp/unit/attribute.h

```diff
@@ -84,6 +91,8 @@ class AttributeContainer {
 		type{t} {}
 
 	attr_type type;
+
+	virtual AttributeContainer *copy() const = 0;
```

Comment:
this is what I had in mind:

    template<typename T>
    T *get_attribute() {
        std::type_info *map_key = &typeid(T)
    
        auto map_idx = attribute_map.find(map_key);
    
        if (map_idx == std::map::end) {
            return nullptr;
        } else {
            return reinterpret_cast<T *>(*map_idx);
        }
    }